### DISCUSSION TITLE
Voltage Supply block on classical controller exmple

### DISCUSSION THREAD

**Discussion by Mario-de-Santi** [upb-lea/gym-electric-motor#246] [Q&A]
Hello, I am using your library to simulate the control of an electric Go-kart for my graduation project.  My idea is to use the "Gym Electric Motor" library and other libraries as an open source alternative for electric vehicle simulations.

I am basing myself on the examples of classic controllers from this repository because I am unable to use your other library, "Gem-Control". I would like to know if by using the examples of classic controllers I can use the Voltage Supply to simulate the current consumption in the batteries of a PMSM motor? (continuous model simulation simulation of the electric Go-kart.)

Moreover, using a PMSM motor, can I use the Finite simulation model? I tried to follow the comments in the "custom_classic_controllers_synch_motor_example.py" program and I am unable to run the simulation.

Code snippets that I altered on Finite model:

    motor_type = 'PMSM'
    control_type = 'SC'
    action_type = 'Finite'

    env_id = action_type + '-' + control_type + '-' + motor_type + '-v0'
    # current_d_controller = {'controller_type': 'pi_controller', 'p_gain': 5, 'i_gain': 250}
    # current_q_controller = {'controller_type': 'pi_controller', 'p_gain': 12, 'i_gain': 250}
    # speed_controller = {'controller_type': 'pi_controller', 'p_gain': 20, 'i_gain': 400}

    current_a_controller = {'controller_type': 'pi_controller', 'p_gain': 5, 'i_gain': 250}
    current_b_controller = {'controller_type': 'pi_controller', 'p_gain': 5, 'i_gain': 250}
    current_c_controller = {'controller_type': 'pi_controller', 'p_gain': 12, 'i_gain': 250}
    speed_controller = {'controller_type': 'pi_controller', 'p_gain': 20, 'i_gain': 400}

    current_controller = [current_a_controller, current_b_controller, current_c_controller]
    overlaid_controller = [speed_controller]

    stages = [current_controller,overlaid_controller]

    controller = Controller.make(env, stages=stages, external_ref_plots=external_ref_plots)

**Comment by max-schenke** [comment]
Hello and thank you for considering GEM.
Indeed, GEM-Control has been integrated entirely into this repository, you can find the latest version on the nightly branch.

Now lets have a look at your questions:

1) Can I use the Voltage Supply to simulate the current consumption in the batteries of a PMSM motor? 

I think you are looking for the DC-link / supply current that flows in between the supply and the inverter? There is currently no default output for this, but you can extract this quantity with the command
`i_sup = env.env.env.physical_system.converter.i_sup(i_abc)`
(Given that `i_abc` containts the momentary values of the abc current in Amperes. Please note that GEM outputs everything in normalized quantities and the outputs must be denormalized before they can be treated as physical quantities.)
The supply models are not very sophisticated yet and I guess you would need to model the battery by other means.

2) Moreover, using a PMSM motor, can I use the Finite simulation model?

Generally, the finite simulation model is ready to use but I have the feeling its not what you want, as it is not compatible with the PI controller.
The finite control set treats the converter legs in terms of switching states ("on" or "off"), whereas the continuous control set allows to directly select the motor phase voltages. Only the latter can make use of the PI controller.

**Reply by Mario-de-Santi** [reply]
Thank you for your attention. I will continue using the continuous model for control simulations and look for other means for battery simulation. Thank you very much.
